### PR TITLE
Fix some typehints in mobject.py

### PR DESCRIPTION
**Pull request by JasonGrace2282**
* Move `typing_extensions` import under `if TYPE_CHECKING`
* Change from using `def animate(self: T ,...) -> T` to `def animate(self, ...) -> Self` as stated in PEP 673
* Fix incorrect usage of `T` in :meth:`.Mobject.set_z_index`
* Change updater return type to `object`. Since we just [discard the result](https://github.com/ManimCommunity/manim/blob/main/manim/mobject/mobject.py#L849-L852), any return value will do.

Note to reviewer:
Please check if this PR breaks autocomplete for the `.animate` syntax.

It seems there is a problem with the `autoaliasattr_directive`, as the `Updater`s are [no longer being shown](https://manimce--3668.org.readthedocs.build/en/3668/reference/manim.mobject.mobject.Mobject.html). Should be fixed after #3671

### CODE REVIEW

File: manim/mobject/mobject.py [hunk 1]
```diff
@@ -237,7 +235,7 @@ def construct(self):
             cls.__init__ = cls._original__init__
 
     @property
-    def animate(self: T) -> _AnimationBuilder | T:
+    def animate(self) -> _AnimationBuilder | Self:
```
Comment (Viicos) [inline manim/mobject/mobject.py:238]: What was the reason to have the union return type in the first place? Is it because `_AnimationBuilder` acts as some kind of proxy object? If so, there's probably a better way to type this, but let's keep this for another PR when this module will be tackled as part of the typing progress

Comment (JasonGrace2282) [reply manim/mobject/mobject.py:238]: > s it because _AnimationBuilder acts as some kind of proxy object?

That's the gist of it!

**Review by behackl — APPROVED** [review]
LGTM

**Comment by JasonGrace2282** [comment]
Note that now #3671 *should be* merged before the next release. Otherwise, as said in the PR description, the type aliases under the `If TYPE_CHECKING` guard won't show up in the docs
